feat: derive Arbitrary implementations

huitseeker opened the pull request:
This is a follow-up to #229 and hence #240 that leans into the derivability of `Arbitrary`. It derives those implementations of `Arbitrary` that trivially can (i.e. those that do not rely on the `FWrap` newtype, or custom behavior such as that of `ScalarStore`).

The regularity of this derivation ensures that these implementations are more maintainable, as we won't have to update them if one of e.g. the Tag enums change.

porcuquine commented on src/scalar_store.rs:
```diff
@@ -216,6 +222,33 @@ pub enum ScalarExpression<F: LurkField> {
+                }
+            ),
+            any::<ScalarThunk<Fr>>().prop_map(|x| Self::Thunk(x)),
+        )
```
Any reason to exclude `Char`, `UInt`, and `Comm`?

porcuquine replied:
Ah, looks like this was just moved from elsewhere…

porcuquine replied:
Unless the goal is to exclude, could we just derive `Arbitrary` for `ScalarExpression`? I do think we might want a custom implementation, so we can support well-formed commitments (i.e. those whose preimage is known). Maybe some percentage could just be random and some well-formed (more expensive, but more useful for some purposes). I'm not saying it needs to happen here, more just trying to understand the implications and thinking through how these things work.

huitseeker replied:
We do derive this now

johnchandlerburnham commented on src/scalar_store.rs:
```diff
@@ -216,6 +222,33 @@ pub enum ScalarExpression<F: LurkField> {
+    type Parameters = ();
+    type Strategy = BoxedStrategy<Self>;
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
```
Minor quibble: in the future we probably don't want to have this Strategy be the default Arbitrary `impl` in favor of something that generates well-formed `ScalarExpression`s. This one here is great for testing serialization/hashing, but not so good for testing parsing/printing/evaluation since the inner `ScalarPtr`s will be noise

johnchandlerburnham commented:
Looks great! Left one comment which is really more of a reminder for the future for us to flex our ability to have multiple strategies for different things

porcuquine commented on src/store.rs:
```diff
@@ -286,6 +288,18 @@ pub struct ScalarPtr<F: LurkField>(F, F);
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        any::<(ExprTag, FWrap<Fr>)>()
+            .prop_map(|(tag, val)| ScalarPtr::from_parts(Fr::from(tag as u64), val.0))
```
Would it make more sense to only create well-formed `ScalarPtr`s — that is, those with valid tags? I guess for testing purposes, it's valuable to have malformed also, but perhaps at a much lower frequency?

Do let me know if I'm misunderstanding or making bad assumptions about how this should work.

porcuquine replied:
For clarity: I do think having a way to generate bad or potentially bad values is generally useful, especially when testing circuits.

huitseeker replied:
> Would it make more sense to only create well-formed ScalarPtrs — that is, those with valid tags? I guess for testing purposes, it's valuable to have malformed also, but perhaps at a much lower frequency?

Yes, naturally.

`Arbitrary` is only one distinguished strategy for creating values of type T. The number of strategies for this type is unbounded, and does not have to involve the `Arbitrary` trait.

In other terms, I can have:
```
struct T {}
impl Arbitrary for T { ... }
fn my_strategy_for_T() -> impl Strategy<Value = T> { ... }
```
and at point of use:
```
proptest! {
  #[test]
  fn my_test_with_arbitrary_T(x in any::<T>()) { // uses Arbitrary
      ... 
  }
  
  #[test]
  fn my_test_with_custom_strategy(x in my_strategy_for_T()) { // does not use Arbitrary
      ....
  }
}
```

porcuquine replied:
Yes, this makes perfect sense and is along the lines of what I ended up thinking we would do. Thank you for the explanation.

porcuquine commented on src/store.rs:
```diff
@@ -375,6 +389,18 @@ pub struct ScalarContPtr<F: LurkField>(F, F);
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        any::<(ContTag, FWrap<Fr>)>()
+            .prop_map(|(tag, val)| ScalarContPtr::from_parts(Fr::from(tag as u64), val.0))
```
Same questions as for `ScalarPtr`.

huitseeker commented:
Thanks for your feedback @johnchandlerburnham @porcuquine. 
I've added the following to the PR:
- a derivation of `Arbitrary` for `ScalarExpression` so that cases of this large enum are not forgotten,
- I've replaced the implementation of `Arbitrary` for `ScalarStore` with a simple custom strategy, as the strategy is clearly marked to be suitable only in some cases.

In general, several strategies for the same type are completely legitimate. We should definitely have more strategies for `Scalar(Cont)?Ptr`, but ideally in upcoming PRs rather than this one.
